### ISSUE TITLE
Bad error message when PyOpenSSL is missing.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4801]
|<img alt="taos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| taos reported|
|-|-|
|Trac ID|trac#4801|
|Type|defect|
|Created|2011-01-12 22:01:24Z|
|Branch|https://github.com/twisted/twisted/tree/msn-ssl-missing-4801|

When a user tries to connect using t.w.p.msn and does not have PyOpenSSL installed there is a type error, producing this exception

```
Email (passport): mypassport@hotmail.com
Password: ******
2009-04-25 10:52:49-0300 [-] Log opened.
2009-04-25 10:52:49-0300 [-] Starting factory <twisted.internet.protocol.ClientFactory instance at 0x9d87e8c>
2009-04-25 10:52:55-0300 [Dispatch,client] Starting factory <twisted.words.protocols.msn.NotificationFactory instance at 0x9e28bcc>
2009-04-25 10:52:55-0300 [Dispatch,client] Stopping factory <twisted.internet.protocol.ClientFactory instance at 0x9d87e8c>
2009-04-25 10:52:55-0300 [Notification,client] Unhandled Error
    Traceback (most recent call last):
      File "/usr/local/lib/python2.5/site-packages/twisted/python/log.py", line 84, in callWithLogger
        return callWithContext({"system": lp}, func, *args, **kw)
      File "/usr/local/lib/python2.5/site-packages/twisted/python/log.py", line 69, in callWithContext
        return context.call({ILogContext: newCtx}, func, *args, **kw)
      File "/usr/local/lib/python2.5/site-packages/twisted/python/context.py", line 59, in callWithContext
        return self.currentContext().callWithContext(ctx, func, *args, **kw)
      File "/usr/local/lib/python2.5/site-packages/twisted/python/context.py", line 37, in callWithContext
        return func(*args,**kw)
    --- <exception caught here> ---
      File "/usr/local/lib/python2.5/site-packages/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
        why = getattr(selectable, method)()
      File "/usr/local/lib/python2.5/site-packages/twisted/internet/tcp.py", line 460, in doRead
        return self.protocol.dataReceived(data)
      File "/usr/local/lib/python2.5/site-packages/twisted/protocols/basic.py", line 238, in dataReceived
        why = self.lineReceived(line)
      File "/usr/local/lib/python2.5/site-packages/twisted/words/protocols/msn.py", line 651, in lineReceived
        handler(params.split())
      File "/usr/local/lib/python2.5/site-packages/twisted/words/protocols/msn.py", line 827, in handle_USR
        d = _login(f.userHandle, f.password, f.passportServer, authData=params[3])
      File "/usr/local/lib/python2.5/site-packages/twisted/words/protocols/msn.py", line 182, in _login
        reactor.connectSSL(_parsePrimitiveHost(nexusServer)[0], 443, fac, ClientContextFactory())
    exceptions.TypeError: 'NoneType' object is not callable

2009-04-25 10:52:55-0300 [Notification,client] Stopping factory <twisted.words.protocols.msn.NotificationFactory instance at 0x9e28bcc>

```
If I am not wrong then this could be fixed via a try and except loop inside the file to detect if it has become a NoneType

Attachments:

* [msn-ClientContextConnection.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5f4/5f4ad80eeb78cb95810f0a82bb91eae5908269fb/53618722d044cee85f0268337b4e4a7011805a50.patch) (2436 bytes) - added by taos on 2011-01-24 11:14:17Z - Fixed with advice.
* [4801.bugfix](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5f4/5f4ad80eeb78cb95810f0a82bb91eae5908269fb/bbe9a5da2b7eaa41d96eaab2fd80e8f9e94ad88c.bugfix) (46 bytes) - added by taos on 2011-01-24 11:43:56Z - NEWS file for my bugfix

<details><summary>Searchable metadata</summary>

```
trac-id__4801 4801
type__defect defect
reporter__taos taos
priority__normal normal
milestone__ 
branch__branches_msn_ssl_missing_4801 branches/msn-ssl-missing-4801
branch_author__exarkun__taos exarkun, taos
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1294869684000000 1294869684000000
changetime__1300244794000000 1300244794000000
version__None None
owner__exarkun exarkun
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="taos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|taos removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to taos|
|-|-|

Thanks Taos.  Some more things to do:

  1. Add a [ReviewProcess#Newsfile news fragment]
  1. Add a unit test
  1. Instead of defining an exception with a `__str__`, just pass the string to the exception's `__init__`.  The base class will do the right thing.
  1. Always use the `raise Foo()` form instead of the `raise Foo` form.
  1. Update the copyright date on `msn.py` to extend to 2011
  1. Keep 3 blank lines between top-level suites (ie add two more blank lines after the new exception class)
  1. Give `PyOpenSSLRequired` a class docstring explaining the case it represents (an operation requiring SSL was attempted but SSL is unavailable due to the missing dependency)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_3):
>   1. Give `PyOpenSSLRequired` a class docstring explaining the case it represents (an operation requiring SSL was attempted but SSL is unavailable due to the missing dependency)

also add a version marker in the docstring: ```@since: 11.0```

Cheers

**Comment by twisted-trac** [comment]
|<img alt="awclin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|awclin removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="awclin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|awclin set owner to taos|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="taos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|taos removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="taos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|taos commented|
|-|-|

Added the unit tests, corrected spacing, changed exception style, added docstrings, updated copyright

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to taos|
|-|-|

 * You should restore the original value of msn.ClientContextFactory after the test (using addCleanup for example).
 * Your test should have a docstring.
 * Only the first 2 arguments of assertRaises are mandatory, you don't have to pass the extra 'None'.
 * The self.exception variable seems pointless, you should pass the exception class directly in assertRaises

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [30933]) Branching to 'msn-ssl-missing-4801'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [30934]) Apply msn-ClientContextConnection.patch

refs [#4801](https://github.com/twisted/twisted/issues/4801)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [30936]) Move the test into an existing TestCase, change it to use NotificationClient instead of _login

refs [#4801](https://github.com/twisted/twisted/issues/4801)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [30937]) Hook up the SSLRequired failure to the loginFailure callback

refs [#4801](https://github.com/twisted/twisted/issues/4801)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun removed owner|
|-|-|

Checked the patch into a branch, made a ... lot ... of changes.  Please review!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_18"></a>|@jesstess set owner to @exarkun|
|-|-|

```
-    for testClass in [PassportTests, NotificationTests,
+    for testClass in [ExceptionTests, PassportTests, NotificationTests,
```

There is no ```ExceptionTests```, although there is a missing ```DispatchTests```.

[builds](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/msn-ssl-missing-4801) are okay (although ```testManyProcesses``` is erroring out ~every run on Windows, on the branch and on trunk).

I think this is good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set status to `closed`|
|-|-|

(In [31165]) Merge msn-ssl-missing-4801

Author: Taos, exarkun
Reviewer: jesstess
Fixes: [#4801](https://github.com/twisted/twisted/issues/4801)


NotificationClient now calls the loginFailure callback when it is unable
to connect to the Passport server due to missing SSL dependencies.
